lib/digest/cryb_sha224.3
.\"-
.\" Copyright (c) 2015 The University of Oslo
.\" Copyright (c) 2016-2022 Dag-Erling Smørgrav
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 15, 2022
.Dt CRYB_SHA224 3
.Os
.Sh NAME
.Nm cryb_sha224_init ,
.Nm cryb_sha224_update ,
.Nm cryb_sha224_final ,
.Nm cryb_sha224_complete
.Nd Secure Hash Algorithm 2
.Sh LIBRARY
.Lb libcryb-digest
.Sh SYNOPSIS
.In stddef.h
.In stdint.h
.In cryb/sha224.h
.Ft void
.Fn cryb_sha224_init "cryb_sha224_ctx *context"
.Ft void
.Fn cryb_sha224_update "cryb_sha224_ctx *context" "const void *data" "size_t len"
.Ft void
.Fn cryb_sha224_final "cryb_sha224_ctx *context" "uint8_t *digest"
.Ft void
.Fn cryb_sha224_complete "const void *data" "size_t len" "uint8_t *digest"
.Sh DESCRIPTION
The
.Nm cryb_sha224
family of functions implements the NIST SHA-224 message digest
algorithm as described in FIPS 180-4.
.Pp
The
.Fn cryb_sha224_init
function initializes the context structure pointed to by
.Va context .
It is the caller's responsibility to allocate this structure.
.Pp
The
.Fn cryb_sha224_update
function hashes the next
.Va len
bytes of data pointed to by the
.Va data
pointer into the given hash context.
.Pp
The
.Fn cryb_sha224_final
function finalizes the computation and writes the resulting message
digest to the caller-provided buffer pointed to by
.Va digest ,
which must be at least
.Dv SHA224_DIGEST_LEN
bytes long.
It is the caller's responsibility to dispose of the context structure
after calling
.Fn cryb_sha224_final .
.Pp
The
.Fn cryb_sha224_complete
function is a shortcut to calling
.Fn cryb_sha224_init ,
.Fn cryb_sha224_update ,
and
.Fn cryb_sha224_final
when the entire message is available up front in a single contiguous
buffer.
.Sh IMPLEMENTATION NOTES
The
.In cryb/sha224.h
header provides macros which allows these functions and the context
type to be referred to without their
.Dq Li cryb_
prefix.
.Sh SEE ALSO
.Xr cryb_digest 3 ,
.Xr cryb_sha1 3 ,
.Xr cryb_sha256 3 ,
.Xr cryb_sha384 3 ,
.Xr cryb_sha512 3
.Sh REFERENCES
.Rs
.%Q National Institute of Standards and Technology
.%R Secure Hash Standard (SHS) (FIPS PUB 180-4)
.%D March 2012
.Re
.Sh AUTHORS
.An -nosplit
These functions and this manual page were written for the University
of Oslo by
.An Dag-Erling Sm\(/orgrav Aq Mt d.e.smorgrav@usit.uio.no .


debian/changelog
cryb-to (0.20190326-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Tue, 26 Mar 2019 12:24:03 +0100

cryb-to (0.20181115-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Thu, 15 Nov 2018 18:58:09 +0100

cryb-to (0.20180513-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Sun, 13 May 2018 16:40:11 +0200

cryb-to (0.20180426-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Thu, 26 Apr 2018 22:04:25 +0200

cryb-to (0.20170904-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Mon, 04 Sep 2017 21:45:06 +0200

cryb-to (0.20170502-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Tue, 02 May 2017 13:16:49 +0200

cryb-to (0.20170406-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 06 Apr 2017 21:29:46 +0200

cryb-to (0.20170322-1) unstable; urgency=low

  * New alpha release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 22 Mar 2017 22:00:35 +0100

cryb-to (0.20170307-1) unstable; urgency=low

  * Initial release

 -- Dag-Erling Smørgrav <des@des.no>  Wed, 09 Mar 2017 22:30:06 +0100


lib/digest/cryb_sha224.c
/*-
 * Copyright (c) 2005-2013 Colin Percival
 * Copyright (c) 2017 Dag-Erling Smørgrav
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "cryb/impl.h"

#include <stdint.h>
#include <string.h>

#include <cryb/bitwise.h>
#include <cryb/endian.h>
#include <cryb/memset_s.h>

#include <cryb/sha224.h>

/* Elementary functions used by SHA224 */
#define Ch(x, y, z)	((x & (y ^ z)) ^ z)
#define Maj(x, y, z)	((x & (y | z)) | (y & z))
#define SHR(x, n)	(x >> n)
#define S0(x)		(ror32(x,  2) ^ ror32(x, 13) ^ ror32(x, 22))
#define S1(x)		(ror32(x,  6) ^ ror32(x, 11) ^ ror32(x, 25))
#define s0(x)		(ror32(x,  7) ^ ror32(x, 18) ^ (x >>  3))
#define s1(x)		(ror32(x, 17) ^ ror32(x, 19) ^ (x >> 10))

/* SHA224 round function */
#define RND(a, b, c, d, e, f, g, h, k)			\
	t0 = h + S1(e) + Ch(e, f, g) + k;		\
	t1 = S0(a) + Maj(a, b, c);			\
	d += t0;					\
	h  = t0 + t1;

/* Adjusted round function for rotating state */
#define RNDr(S, W, i, k)			\
	RND(S[(64 - i) % 8], S[(65 - i) % 8],	\
	    S[(66 - i) % 8], S[(67 - i) % 8],	\
	    S[(68 - i) % 8], S[(69 - i) % 8],	\
	    S[(70 - i) % 8], S[(71 - i) % 8],	\
	    W[i] + k)

/*
 * SHA224 block compression function.  The 256-bit state is transformed via
 * the 512-bit input block to produce a new state.
 */
static void
sha224_Transform(uint32_t *state, const uint8_t block[64])
{
	uint32_t W[64];
	uint32_t S[8];
	uint32_t t0, t1;
	int i;

	/* 1. Prepare message schedule W. */
	be32decv(W, block, 16);
	for (i = 16; i < 64; i++)
		W[i] = s1(W[i - 2]) + W[i - 7] + s0(W[i - 15]) + W[i - 16];

	/* 2. Initialize working variables. */
	memcpy(S, state, 32);

	/* 3. Mix. */
	RNDr(S, W,  0, 0x428a2f98);
	RNDr(S, W,  1, 0x71374491);
	RNDr(S, W,  2, 0xb5c0fbcf);
	RNDr(S, W,  3, 0xe9b5dba5);
	RNDr(S, W,  4, 0x3956c25b);
	RNDr(S, W,  5, 0x59f111f1);
	RNDr(S, W,  6, 0x923f82a4);
	RNDr(S, W,  7, 0xab1c5ed5);
	RNDr(S, W,  8, 0xd807aa98);
	RNDr(S, W,  9, 0x12835b01);
	RNDr(S, W, 10, 0x243185be);
	RNDr(S, W, 11, 0x550c7dc3);
	RNDr(S, W, 12, 0x72be5d74);
	RNDr(S, W, 13, 0x80deb1fe);
	RNDr(S, W, 14, 0x9bdc06a7);
	RNDr(S, W, 15, 0xc19bf174);
	RNDr(S, W, 16, 0xe49b69c1);
	RNDr(S, W, 17, 0xefbe4786);
	RNDr(S, W, 18, 0x0fc19dc6);
	RNDr(S, W, 19, 0x240ca1cc);
	RNDr(S, W, 20, 0x2de92c6f);
	RNDr(S, W, 21, 0x4a7484aa);
	RNDr(S, W, 22, 0x5cb0a9dc);
	RNDr(S, W, 23, 0x76f988da);
	RNDr(S, W, 24, 0x983e5152);
	RNDr(S, W, 25, 0xa831c66d);
	RNDr(S, W, 26, 0xb00327c8);
	RNDr(S, W, 27, 0xbf597fc7);
	RNDr(S, W, 28, 0xc6e00bf3);
	RNDr(S, W, 29, 0xd5a79147);
	RNDr(S, W, 30, 0x06ca6351);
	RNDr(S, W, 31, 0x14292967);
	RNDr(S, W, 32, 0x27b70a85);
	RNDr(S, W, 33, 0x2e1b2138);
	RNDr(S, W, 34, 0x4d2c6dfc);
	RNDr(S, W, 35, 0x53380d13);
	RNDr(S, W, 36, 0x650a7354);
	RNDr(S, W, 37, 0x766a0abb);
	RNDr(S, W, 38, 0x81c2c92e);
	RNDr(S, W, 39, 0x92722c85);
	RNDr(S, W, 40, 0xa2bfe8a1);
	RNDr(S, W, 41, 0xa81a664b);
	RNDr(S, W, 42, 0xc24b8b70);
	RNDr(S, W, 43, 0xc76c51a3);
	RNDr(S, W, 44, 0xd192e819);
	RNDr(S, W, 45, 0xd6990624);
	RNDr(S, W, 46, 0xf40e3585);
	RNDr(S, W, 47, 0x106aa070);
	RNDr(S, W, 48, 0x19a4c116);
	RNDr(S, W, 49, 0x1e376c08);
	RNDr(S, W, 50, 0x2748774c);
	RNDr(S, W, 51, 0x34b0bcb5);
	RNDr(S, W, 52, 0x391c0cb3);
	RNDr(S, W, 53, 0x4ed8aa4a);
	RNDr(S, W, 54, 0x5b9cca4f);
	RNDr(S, W, 55, 0x682e6ff3);
	RNDr(S, W, 56, 0x748f82ee);
	RNDr(S, W, 57, 0x78a5636f);
	RNDr(S, W, 58, 0x84c87814);
	RNDr(S, W, 59, 0x8cc70208);
	RNDr(S, W, 60, 0x90befffa);
	RNDr(S, W, 61, 0xa4506ceb);
	RNDr(S, W, 62, 0xbef9a3f7);
	RNDr(S, W, 63, 0xc67178f2);

	/* 4. Mix local working variables into global state. */
	for (i = 0; i < 8; i++)
		state[i] += S[i];
}

static uint8_t PAD[64] = {
	0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
};

/* Add padding and terminating bit-count. */
static void
sha224_pad(sha224_ctx *ctx)
{
	uint8_t len[8];
	uint32_t r, plen;

	/*
	 * Convert length to a vector of bytes -- we do this now rather
	 * than later because the length will change after we pad.
	 */
	be64enc(len, ctx->count);

	/* Add 1--64 bytes so that the resulting length is 56 mod 64. */
	r = (ctx->count >> 3) & 0x3f;
	plen = (r < 56) ? (56 - r) : (120 - r);
	sha224_update(ctx, PAD, (size_t)plen);

	/* Add the terminating bit-count. */
	sha224_update(ctx, len, 8);
}

/**
 * sha224_init(ctx):
 * Initialize the SHA224 context ${ctx}.
 */
void
sha224_init(sha224_ctx *ctx)
{

	/* Zero bits processed so far. */
	ctx->count = 0;

	/* Magic initialization constants. */
	ctx->state[0] = 0xc1059ed8;
	ctx->state[1] = 0x367cd507;
	ctx->state[2] = 0x3070dd17;
	ctx->state[3] = 0xf70e5939;
	ctx->state[4] = 0xffc00b31;
	ctx->state[5] = 0x68581511;
	ctx->state[6] = 0x64f98fa7;
	ctx->state[7] = 0xbefa4fa4;
}

/**
 * sha224_update(ctx, in, len):
 * Input ${len} bytes from ${in} into the SHA224 context ${ctx}.
 */
void
sha224_update(sha224_ctx *ctx, const void *in, size_t len)
{
	uint32_t r;
	const uint8_t *src = in;

	/* Return immediately if we have nothing to do. */
	if (len == 0)
		return;

	/* Number of bytes left in the buffer from previous updates. */
	r = (ctx->count >> 3) & 0x3f;

	/* Update number of bits. */
	ctx->count += (uint64_t)(len) << 3;

	/* Handle the case where we don't need to perform any transforms. */
	if (len < 64 - r) {
		memcpy(&ctx->buf[r], src, len);
		return;
	}

	/* Finish the current block. */
	memcpy(&ctx->buf[r], src, 64 - r);
	sha224_Transform(ctx->state, ctx->buf);
	src += 64 - r;
	len -= 64 - r;

	/* Perform complete blocks. */
	while (len >= 64) {
		sha224_Transform(ctx->state, src);
		src += 64;
		len -= 64;
	}

	/* Copy left over data into buffer. */
	memcpy(ctx->buf, src, len);
}

/**
 * sha224_final(ctx, digest):
 * Output the SHA224 hash of the data input to the context ${ctx} into the
 * buffer ${digest}.
 */
void
sha224_final(sha224_ctx *ctx, uint8_t *digest)
{

	/* Add padding. */
	sha224_pad(ctx);

	/* Write the hash. */
	be32encv(digest, ctx->state, SHA224_DIGEST_LEN / 4);

	/* Clear the context state. */
	memset_s(ctx, 0, sizeof *ctx, sizeof *ctx);
}

/**
 * sha224_complete(in, len, digest):
 * Compute the SHA224 hash of ${len} bytes from $in} and write it to ${digest}.
 */
void
sha224_complete(const void *in, size_t len, uint8_t *digest)
{
	sha224_ctx ctx;

	sha224_init(&ctx);
	sha224_update(&ctx, in, len);
	sha224_final(&ctx, digest);
}

digest_algorithm sha224_digest = {
	.name			 = "sha224",
	.contextlen		 = sizeof(sha224_ctx),
	.blocklen		 = SHA224_BLOCK_LEN,
	.digestlen		 = SHA224_DIGEST_LEN,
	.init			 = (digest_init_func)(void *)sha224_init,
	.update			 = (digest_update_func)(void *)sha224_update,
	.final			 = (digest_final_func)(void *)sha224_final,
	.complete		 = (digest_complete_func)(void *)sha224_complete,
};
